### PR TITLE
[PXP-2270] Delete user

### PR DESCRIPTION
**Pull request by vpsx**
### New Features
- Set delete cascades in Fence db in general: now deleting a row in a parent table will delete the appropriate child rows too.
- Add delete-user endpoint to remove user and user-related data, both from Fence db and from Google if applicable.

### Dependency updates
- bump cirrus
- bump userdatamodel

### CODE REVIEW

**Comment by PlanXCyborg** [comment]
This PR contains code that is not formatted correctly according to [`black`](https://github.com/ambv/black). Run `black` on your code before merging.

<details>
<summary>Expand the full diff to see formatting changes</summary>

```diff
--- fence/models.py
+++ blackened
@@ -104,18 +104,14 @@
 
     # human readable description, not required
     description = Column(String(400))
 
     # required if you need to support client credential
-    user_id = Column(Integer, ForeignKey(User.id, ondelete='CASCADE'))
+    user_id = Column(Integer, ForeignKey(User.id, ondelete="CASCADE"))
     user = relationship(
         "User",
-        backref=backref(
-            "clients",
-            cascade="all, delete-orphan",
-            passive_deletes=True,
-        ),
+        backref=backref("clients", cascade="all, delete-orphan", passive_deletes=True),
     )
 
     # this is for internal microservices to skip user grant
     auto_approve = Column(Boolean, default=False)
 
@@ -241,13 +237,11 @@
 
     user_id = Column(Integer, ForeignKey("User.id", ondelete="CASCADE"))
     user = relationship(
         "User",
         backref=backref(
-            "authorization_codes",
-            cascade="all, delete-orphan",
-            passive_deletes=True,
+            "authorization_codes", cascade="all, delete-orphan", passive_deletes=True
         ),
     )
 
     nonce = Column(String, nullable=True)
 
@@ -290,11 +284,11 @@
     # case we're ever juggling mult. projects)
     google_unique_id = Column(String, unique=False, nullable=False)
 
     client_id = Column(String(40))
 
-    user_id = Column(Integer, ForeignKey(User.id, ondelete='CASCADE'), nullable=False)
+    user_id = Column(Integer, ForeignKey(User.id, ondelete="CASCADE"), nullable=False)
     user = relationship(
         "User",
         backref=backref(
             "google_service_accounts",
             cascade="all, delete-orphan",
@@ -318,20 +312,17 @@
 
     id = Column(Integer, primary_key=True)
 
     email = Column(String, unique=True, nullable=False)
 
-    user_id = Column(Integer, ForeignKey(User.id, ondelete='CASCADE'), nullable=False)
+    user_id = Column(Integer, ForeignKey(User.id, ondelete="CASCADE"), nullable=False)
     user = relationship(
         "User",
         backref=backref(
-            "user_google_accounts",
-            cascade="all, delete-orphan",
-            passive_deletes=True,
-        ),
-    )
-
+            "user_google_accounts", cascade="all, delete-orphan", passive_deletes=True
+        ),
+    )
 
     def delete(self):
         with flask.current_app.db.session as session:
             session.delete(self)
             session.commit()
@@ -340,23 +331,37 @@
 
 class UserGoogleAccountToProxyGroup(Base):
     __tablename__ = "user_google_account_to_proxy_group"
 
     user_google_account_id = Column(
-        Integer, ForeignKey(UserGoogleAccount.id, ondelete='CASCADE'), nullable=False, primary_key=True
+        Integer,
+        ForeignKey(UserGoogleAccount.id, ondelete="CASCADE"),
+        nullable=False,
+        primary_key=True,
     )
     user_google_account = relationship(
-            "UserGoogleAccount",
-            backref=backref("user_google_account_to_proxy_group", cascade="all, delete-orphan", passive_deletes=True),
+        "UserGoogleAccount",
+        backref=backref(
+            "user_google_account_to_proxy_group",
+            cascade="all, delete-orphan",
+            passive_deletes=True,
+        ),
     )
 
     proxy_group_id = Column(
-        String, ForeignKey(GoogleProxyGroup.id, ondelete='CASCADE'), nullable=False, primary_key=True
+        String,
+        ForeignKey(GoogleProxyGroup.id, ondelete="CASCADE"),
+        nullable=False,
+        primary_key=True,
     )
     google_proxy_group = relationship(
-            "GoogleProxyGroup",
-            backref=backref("user_google_account_to_proxy_group", cascade="all, delete-orphan", passive_deletes=True),
+        "GoogleProxyGroup",
+        backref=backref(
+            "user_google_account_to_proxy_group",
+            cascade="all, delete-orphan",
+            passive_deletes=True,
+        ),
     )
 
     expires = Column(BigInteger)
 
     def delete(self):
@@ -372,15 +377,19 @@
     id = Column(Integer, primary_key=True)
 
     key_id = Column(String, nullable=False)
 
     service_account_id = Column(
-        Integer, ForeignKey(GoogleServiceAccount.id, ondelete='CASCADE'), nullable=False
+        Integer, ForeignKey(GoogleServiceAccount.id, ondelete="CASCADE"), nullable=False
     )
     google_service_account = relationship(
-            "GoogleServiceAccount",
-            backref=backref("google_service_account_keys", cascade="all, delete-orphan", passive_deletes=True),
+        "GoogleServiceAccount",
+        backref=backref(
+            "google_service_account_keys",
+            cascade="all, delete-orphan",
+            passive_deletes=True,
+        ),
     )
 
     expires = Column(BigInteger)
 
     private_key = Column(String)
@@ -394,18 +403,20 @@
 
 class GoogleBucketAccessGroup(Base):
     __tablename__ = "google_bucket_access_group"
     id = Column(Integer, primary_key=True)
 
-    bucket_id = Column(Integer, ForeignKey(Bucket.id, ondelete='CASCADE'), nullable=False)
+    bucket_id = Column(
+        Integer, ForeignKey(Bucket.id, ondelete="CASCADE"), nullable=False
+    )
     bucket = relationship(
         "Bucket",
         backref=backref(
-                    "google_bucket_access_groups",
-                    cascade="all, delete-orphan",
-                    passive_deletes=True,
-                ),
+            "google_bucket_access_groups",
+            cascade="all, delete-orphan",
+            passive_deletes=True,
+        ),
     )
 
     email = Column(String, nullable=False)
 
     # specify what kind of storage access this group has e.g. ['read-storage']
@@ -420,30 +431,34 @@
 
 class GoogleProxyGroupToGoogleBucketAccessGroup(Base):
     __tablename__ = "google_proxy_group_to_google_bucket_access_group"
     id = Column(Integer, primary_key=True)
 
-    proxy_group_id = Column(String, ForeignKey(GoogleProxyGroup.id, ondelete='CASCADE'), nullable=False)
+    proxy_group_id = Column(
+        String, ForeignKey(GoogleProxyGroup.id, ondelete="CASCADE"), nullable=False
+    )
     proxy_group = relationship(
         "GoogleProxyGroup",
         backref=backref(
-                    "google_proxy_group_to_google_bucket_access_group",
-                    cascade="all, delete-orphan",
-                    passive_deletes=True,
-                ),
+            "google_proxy_group_to_google_bucket_access_group",
+            cascade="all, delete-orphan",
+            passive_deletes=True,
+        ),
     )
 
     access_group_id = Column(
-        Integer, ForeignKey(GoogleBucketAccessGroup.id, ondelete='CASCADE'), nullable=False
+        Integer,
+        ForeignKey(GoogleBucketAccessGroup.id, ondelete="CASCADE"),
+        nullable=False,
     )
     access_group = relationship(
         "GoogleBucketAccessGroup",
         backref=backref(
-                "google_proxy_group_to_google_bucket_access_group",
-                cascade="all, delete-orphan",
-                passive_deletes=True,
-            ),
+            "google_proxy_group_to_google_bucket_access_group",
+            cascade="all, delete-orphan",
+            passive_deletes=True,
+        ),
     )
 
 
 class UserServiceAccount(Base):
     __tablename__ = "user_service_account"
@@ -462,61 +477,57 @@
 class ServiceAccountAccessPrivilege(Base):
     __tablename__ = "service_account_access_privilege"
 
     id = Column(Integer, primary_key=True)
 
-    project_id = Column(Integer, ForeignKey(Project.id, ondelete='CASCADE'), nullable=False)
+    project_id = Column(
+        Integer, ForeignKey(Project.id, ondelete="CASCADE"), nullable=False
+    )
     project = relationship(
         "Project",
         backref=backref(
-            "sa_access_privileges",
-            cascade="all, delete-orphan",
-            passive_deletes=True,
+            "sa_access_privileges", cascade="all, delete-orphan", passive_deletes=True
         ),
     )
 
     service_account_id = Column(
-        Integer, ForeignKey(UserServiceAccount.id, ondelete='CASCADE'), nullable=False
+        Integer, ForeignKey(UserServiceAccount.id, ondelete="CASCADE"), nullable=False
     )
     service_account = relationship(
         "UserServiceAccount",
         backref=backref(
-            "access_privileges",
-            cascade="all, delete-orphan",
-            passive_deletes=True,
+            "access_privileges", cascade="all, delete-orphan", passive_deletes=True
         ),
     )
 
 
 class ServiceAccountToGoogleBucketAccessGroup(Base):
     __tablename__ = "service_account_to_google_bucket_access_group"
     id = Column(Integer, primary_key=True)
 
     service_account_id = Column(
-        Integer, ForeignKey(UserServiceAccount.id, ondelete='CASCADE'), nullable=False
+        Integer, ForeignKey(UserServiceAccount.id, ondelete="CASCADE"), nullable=False
     )
     service_account = relationship(
         "UserServiceAccount",
         backref=backref(
-            "to_access_groups",
-            cascade="all, delete-orphan",
-            passive_deletes=True,
+            "to_access_groups", cascade="all, delete-orphan", passive_deletes=True
         ),
     )
 
     expires = Column(BigInteger)
 
     access_group_id = Column(
-        Integer, ForeignKey(GoogleBucketAccessGroup.id, ondelete='CASCADE'), nullable=False
+        Integer,
+        ForeignKey(GoogleBucketAccessGroup.id, ondelete="CASCADE"),
+        nullable=False,
     )
 
     access_group = relationship(
         "GoogleBucketAccessGroup",
         backref=backref(
-            "to_access_groups",
-            cascade="all, delete-orphan",
-            passive_deletes=True,
+            "to_access_groups", cascade="all, delete-orphan", passive_deletes=True
         ),
     )
 
 
 to_timestamp = (
@@ -717,19 +728,31 @@
 
 def set_foreign_key_constraint_on_delete_cascade(
     table_name, column_name, fk_table_name, fk_column_name, driver, metadata
 ):
     set_foreign_key_constraint_on_delete(
-        table_name, column_name, fk_table_name, fk_column_name, "CASCADE", driver, metadata
+        table_name,
+        column_name,
+        fk_table_name,
+        fk_column_name,
+        "CASCADE",
+        driver,
+        metadata,
     )
 
 
 def set_foreign_key_constraint_on_delete_setnull(
     table_name, column_name, fk_table_name, fk_column_name, driver, metadata
 ):
     set_foreign_key_constraint_on_delete(
-        table_name, column_name, fk_table_name, fk_column_name, "SET NULL", driver, metadata
+        table_name,
+        column_name,
+        fk_table_name,
+        fk_column_name,
+        "SET NULL",
+        driver,
+        metadata,
     )
 
 
 def set_foreign_key_constraint_on_delete(
     table_name, column_name, fk_table_name, fk_column_name, ondelete, driver, metadata
@@ -737,11 +760,12 @@
     table = Table(table_name, metadata, autoload=True, autoload_with=driver.engine)
     foreign_key_name = "{}_{}_fkey".format(table_name.lower(), column_name)
 
     if column_name in table.c:
         import pdb
-        #pdb.set_trace()
+
+        # pdb.set_trace()
         print("About to start execute...")
         foreign_keys = [fk.name for fk in getattr(table.c, column_name).foreign_keys]
         print("PRINTING FOREIGN KEYS IN SET FUNCTION......")
         print(foreign_keys)
         with driver.session as session:
@@ -976,119 +1000,159 @@
         session.commit()
 
 
 def _set_on_delete_cascades(driver, md):
     set_foreign_key_constraint_on_delete_cascade(
-        'client', 'user_id', 'User', 'id', driver, md
-    )
-    set_foreign_key_constraint_on_delete_cascade(
-        'authorization_code', 'user_id', 'User', 'id', driver, md
-    )
-    set_foreign_key_constraint_on_delete_cascade(
-        'google_service_account', 'user_id', 'User', 'id', driver, md
-    )
-    set_foreign_key_constraint_on_delete_cascade(
-        'user_google_account', 'user_id', 'User', 'id', driver, md
-    )
-    set_foreign_key_constraint_on_delete_cascade(
-        'user_google_account_to_proxy_group', 'user_google_account_id', 'user_google_account', 'id', driver, md
-    )
-    set_foreign_key_constraint_on_delete_cascade(
-        'user_google_account_to_proxy_group', 'proxy_group_id', 'google_proxy_group', 'id', driver, md
-    )
-    set_foreign_key_constraint_on_delete_cascade(
-        'google_service_account_key', 'service_account_id', 'google_service_account', 'id', driver, md
-    )
-    set_foreign_key_constraint_on_delete_cascade(
-        'google_bucket_access_group', 'bucket_id', 'bucket', 'id', driver, md
-    )
-    set_foreign_key_constraint_on_delete_cascade(
-        'google_proxy_group_to_google_bucket_access_group', 'proxy_group_id', 'google_proxy_group', 'id', driver, md
-    )
-    set_foreign_key_constraint_on_delete_cascade(
-        'google_proxy_group_to_google_bucket_access_group', 'access_group_id', 'google_bucket_access_group', 'id', driver, md
-    )
-    set_foreign_key_constraint_on_delete_cascade(
-        'service_account_access_privilege', 'project_id', 'project', 'id', driver, md
-    )
-    set_foreign_key_constraint_on_delete_cascade(
-        'service_account_access_privilege', 'service_account_id', 'user_service_account', 'id', driver, md
-    )
-    set_foreign_key_constraint_on_delete_cascade(
-        'service_account_to_google_bucket_access_group', 'service_account_id', 'user_service_account', 'id', driver, md
-    )
-    set_foreign_key_constraint_on_delete_cascade(
-        'service_account_to_google_bucket_access_group', 'access_group_id', 'google_bucket_access_group', 'id', driver, md
-    )
-    set_foreign_key_constraint_on_delete_cascade(
-        'hmac_keypair', 'user_id', 'User', 'id', driver, md
-    )
-    set_foreign_key_constraint_on_delete_cascade(
-        'hmac_keypair_archive', 'user_id', 'User', 'id', driver, md
-    )
-    set_foreign_key_constraint_on_delete_cascade(
-        'user_to_group', 'user_id', 'User', 'id', driver, md
-    )
-    set_foreign_key_constraint_on_delete_cascade(
-        'user_to_group', 'group_id', 'Group', 'id', driver, md
-    )
-    set_foreign_key_constraint_on_delete_cascade(
-        'access_privilege', 'user_id', 'User', 'id', driver, md
-    )
-    set_foreign_key_constraint_on_delete_cascade(
-        'access_privilege', 'group_id', 'Group', 'id', driver, md
-    )
-    set_foreign_key_constraint_on_delete_cascade(
-        'access_privilege', 'project_id', 'project', 'id', driver, md
-    )
-    set_foreign_key_constraint_on_delete_cascade(
-        'access_privilege', 'provider_id', 'authorization_provider', 'id', driver, md
-    )
-    set_foreign_key_constraint_on_delete_cascade(
-        'user_to_bucket', 'user_id', 'User', 'id', driver, md
-    )
-    set_foreign_key_constraint_on_delete_cascade(
-        'user_to_bucket', 'bucket_id', 'bucket', 'id', driver, md
-    )
-    set_foreign_key_constraint_on_delete_cascade(
-        'bucket', 'provider_id', 'cloud_provider', 'id', driver, md
-    )
-    set_foreign_key_constraint_on_delete_cascade(
-        'project_to_bucket', 'project_id', 'project', 'id', driver, md
-    )
-    set_foreign_key_constraint_on_delete_cascade(
-        'project_to_bucket', 'bucket_id', 'bucket', 'id', driver, md
-    )
-    set_foreign_key_constraint_on_delete_cascade(
-        'compute_access', 'project_id', 'project', 'id', driver, md
-    )
-    set_foreign_key_constraint_on_delete_cascade(
-        'compute_access', 'user_id', 'User', 'id', driver, md
-    )
-    set_foreign_key_constraint_on_delete_cascade(
-        'compute_access', 'group_id', 'Group', 'id', driver, md
-    )
-    set_foreign_key_constraint_on_delete_cascade(
-        'compute_access', 'provider_id', 'cloud_provider', 'id', driver, md
-    )
-    set_foreign_key_constraint_on_delete_cascade(
-        'storage_access', 'project_id', 'project', 'id', driver, md
-    )
-    set_foreign_key_constraint_on_delete_cascade(
-        'storage_access', 'user_id', 'User', 'id', driver, md
-    )
-    set_foreign_key_constraint_on_delete_cascade(
-        'storage_access', 'group_id', 'Group', 'id', driver, md
-    )
-    set_foreign_key_constraint_on_delete_cascade(
-        'storage_access', 'provider_id', 'cloud_provider', 'id', driver, md
-    )
-    set_foreign_key_constraint_on_delete_cascade(
-        'certificate', 'application_id', 'application', 'id', driver, md
-    )
-    set_foreign_key_constraint_on_delete_cascade(
-        's3credential', 'user_id', 'User', 'id', driver, md
-    )
-    set_foreign_key_constraint_on_delete_cascade(
-        'tag', 'user_id', 'User', 'id', driver, md
-    )
-
+        "client", "user_id", "User", "id", driver, md
+    )
+    set_foreign_key_constraint_on_delete_cascade(
+        "authorization_code", "user_id", "User", "id", driver, md
+    )
+    set_foreign_key_constraint_on_delete_cascade(
+        "google_service_account", "user_id", "User", "id", driver, md
+    )
+    set_foreign_key_constraint_on_delete_cascade(
+        "user_google_account", "user_id", "User", "id", driver, md
+    )
+    set_foreign_key_constraint_on_delete_cascade(
+        "user_google_account_to_proxy_group",
+        "user_google_account_id",
+        "user_google_account",
+        "id",
+        driver,
+        md,
+    )
+    set_foreign_key_constraint_on_delete_cascade(
+        "user_google_account_to_proxy_group",
+        "proxy_group_id",
+        "google_proxy_group",
+        "id",
+        driver,
+        md,
+    )
+    set_foreign_key_constraint_on_delete_cascade(
+        "google_service_account_key",
+        "service_account_id",
+        "google_service_account",
+        "id",
+        driver,
+        md,
+    )
+    set_foreign_key_constraint_on_delete_cascade(
+        "google_bucket_access_group", "bucket_id", "bucket", "id", driver, md
+    )
+    set_foreign_key_constraint_on_delete_cascade(
+        "google_proxy_group_to_google_bucket_access_group",
+        "proxy_group_id",
+        "google_proxy_group",
+        "id",
+        driver,
+        md,
+    )
+    set_foreign_key_constraint_on_delete_cascade(
+        "google_proxy_group_to_google_bucket_access_group",
+        "access_group_id",
+        "google_bucket_access_group",
+        "id",
+        driver,
+        md,
+    )
+    set_foreign_key_constraint_on_delete_cascade(
+        "service_account_access_privilege", "project_id", "project", "id", driver, md
+    )
+    set_foreign_key_constraint_on_delete_cascade(
+        "service_account_access_privilege",
+        "service_account_id",
+        "user_service_account",
+        "id",
+        driver,
+        md,
+    )
+    set_foreign_key_constraint_on_delete_cascade(
+        "service_account_to_google_bucket_access_group",
+        "service_account_id",
+        "user_service_account",
+        "id",
+        driver,
+        md,
+    )
+    set_foreign_key_constraint_on_delete_cascade(
+        "service_account_to_google_bucket_access_group",
+        "access_group_id",
+        "google_bucket_access_group",
+        "id",
+        driver,
+        md,
+    )
+    set_foreign_key_constraint_on_delete_cascade(
+        "hmac_keypair", "user_id", "User", "id", driver, md
+    )
+    set_foreign_key_constraint_on_delete_cascade(
+        "hmac_keypair_archive", "user_id", "User", "id", driver, md
+    )
+    set_foreign_key_constraint_on_delete_cascade(
+        "user_to_group", "user_id", "User", "id", driver, md
+    )
+    set_foreign_key_constraint_on_delete_cascade(
+        "user_to_group", "group_id", "Group", "id", driver, md
+    )
+    set_foreign_key_constraint_on_delete_cascade(
+        "access_privilege", "user_id", "User", "id", driver, md
+    )
+    set_foreign_key_constraint_on_delete_cascade(
+        "access_privilege", "group_id", "Group", "id", driver, md
+    )
+    set_foreign_key_constraint_on_delete_cascade(
+        "access_privilege", "project_id", "project", "id", driver, md
+    )
+    set_foreign_key_constraint_on_delete_cascade(
+        "access_privilege", "provider_id", "authorization_provider", "id", driver, md
+    )
+    set_foreign_key_constraint_on_delete_cascade(
+        "user_to_bucket", "user_id", "User", "id", driver, md
+    )
+    set_foreign_key_constraint_on_delete_cascade(
+        "user_to_bucket", "bucket_id", "bucket", "id", driver, md
+    )
+    set_foreign_key_constraint_on_delete_cascade(
+        "bucket", "provider_id", "cloud_provider", "id", driver, md
+    )
+    set_foreign_key_constraint_on_delete_cascade(
+        "project_to_bucket", "project_id", "project", "id", driver, md
+    )
+    set_foreign_key_constraint_on_delete_cascade(
+        "project_to_bucket", "bucket_id", "bucket", "id", driver, md
+    )
+    set_foreign_key_constraint_on_delete_cascade(
+        "compute_access", "project_id", "project", "id", driver, md
+    )
+    set_foreign_key_constraint_on_delete_cascade(
+        "compute_access", "user_id", "User", "id", driver, md
+    )
+    set_foreign_key_constraint_on_delete_cascade(
+        "compute_access", "group_id", "Group", "id", driver, md
+    )
+    set_foreign_key_constraint_on_delete_cascade(
+        "compute_access", "provider_id", "cloud_provider", "id", driver, md
+    )
+    set_foreign_key_constraint_on_delete_cascade(
+        "storage_access", "project_id", "project", "id", driver, md
+    )
+    set_foreign_key_constraint_on_delete_cascade(
+        "storage_access", "user_id", "User", "id", driver, md
+    )
+    set_foreign_key_constraint_on_delete_cascade(
+        "storage_access", "group_id", "Group", "id", driver, md
+    )
+    set_foreign_key_constraint_on_delete_cascade(
+        "storage_access", "provider_id", "cloud_provider", "id", driver, md
+    )
+    set_foreign_key_constraint_on_delete_cascade(
+        "certificate", "application_id", "application", "id", driver, md
+    )
+    set_foreign_key_constraint_on_delete_cascade(
+        "s3credential", "user_id", "User", "id", driver, md
+    )
+    set_foreign_key_constraint_on_delete_cascade(
+        "tag", "user_id", "User", "id", driver, md
+    )
+

--- fence/resources/admin/admin_users.py
+++ blackened
@@ -221,11 +221,13 @@
             )
             pgname = _get_proxy_group_name_for_user(
                 user.id, user.username, prefix=config["GOOGLE_GROUP_PREFIX"]
             )
             google_proxy_group_g = gcm.get_group(pgname)
-            gpg_email = google_proxy_group_g.get("email") if google_proxy_group_g else None
+            gpg_email = (
+                google_proxy_group_g.get("email") if google_proxy_group_g else None
+            )
 
         if not gpg_email:
             capp.logger.info(
                 "Could not find Google proxy group for user in Fence db or in cirrus. "
                 "Assuming Google not in use as IdP. Proceeding with Fence deletes."
@@ -243,13 +245,11 @@
                 # Upon deletion of a service account, Google will
                 # automatically delete all key IDs associated with that
                 # service account. So we skip doing that here.
                 capp.logger.debug(
                     "Attempting to delete Google service account with email {} "
-                    "along with all associated service account keys...".format(
-                        sae
-                    )
+                    "along with all associated service account keys...".format(sae)
                 )
                 r = gcm.delete_service_account(sae)
 
                 if r == {}:
                     capp.logger.info(
@@ -283,13 +283,11 @@
                             )
                         )
                     else:
                         capp.logger.info(
                             "Records for service account {} NOT FOUND in Fence database. "
-                            "Continuing anyway.".format(
-                                sae
-                            )
+                            "Continuing anyway.".format(sae)
                         )
 
                 else:
                     raise UnavailableError(
                         "Error: Google unable to delete service account {}. Aborting".format(
```
</details></br>

This formatting comment was generated automatically by a script in [uc-cdis/wool](https://github.com/uc-cdis/wool).

File: fence/resources/admin/admin_users.py [hunk 1]
```diff
@@ -150,19 +162,236 @@ def add_user_to_projects(current_session, username, projects=None):
...
+    r = gcm.delete_group(gpg_email)
+
+    if r == {}:
+        capp.logger.info(
+            "Google proxy group with email {} successfully removed from Google.".format(
+                gpg_email
+            )
+        )
+        if google_proxy_group_f:
+            # (else it was google_proxy_group_*g* and there is nothing to delete in Fence db.)
```
Comment (Avantol13) [inline fence/resources/admin/admin_users.py]: what does the *g* mean?

File: fence/resources/admin/admin_users.py [hunk 2]
```diff
@@ -150,19 +162,236 @@ def add_user_to_projects(current_session, username, projects=None):
...
+            capp.logger.debug("Attempting to clear records from Fence database...")
+            capp.logger.debug(
+                "Deleting rows in google_proxy_group_to_google_bucket_access_group..."
```
Comment (Avantol13) [inline fence/resources/admin/admin_users.py]: you could be fancy and programatically get the tablename with the `Class.__tablename__` but we're probably never going to change it so not a big deal

File: fence/resources/admin/admin_users.py [hunk 3]
```diff
@@ -150,19 +162,236 @@ def add_user_to_projects(current_session, username, projects=None):
...
-        return {"result": "success"}
+    capp.logger.debug("Beginning delete user.")
+
+    with GoogleCloudManager() as gcm:
+        # Delete user's service accounts, SA keys, user proxy group from Google.
+        # Noop if Google not in use.
+
+        user = query_for_user(session=current_session, username=username)
+        if not user:
+            raise NotFound("".join(["user name ", username, " not found"]))
```
Comment (Avantol13) [inline fence/resources/admin/admin_users.py]:
```suggestion
            raise NotFound("user name {} not found".format(username))
```

I find this just a little easier to read but it's up to you

File: fence/resources/admin/admin_users.py [hunk 4]
```diff
@@ -1,5 +1,17 @@
-from fence.errors import NotFound, UserError
-from fence.models import User
+from cirrus import GoogleCloudManager
+from cirrus.google_cloud.manager import _get_proxy_group_name_for_user
```
Comment (Avantol13) [inline fence/resources/admin/admin_users.py]: we should do a quick PR to make this not have the leading underscore, since we're no longer just using it internally to cirrus but opening it to public

File: tests/admin/test_admin_users.py [hunk 5]
```diff
@@ -26,6 +26,9 @@ def test_create_user(db_session, oauth_client):
     assert user.email == "insert_user@fake.com"
 
 
+@pytest.mark.skip(
```
Comment (Avantol13) [inline tests/admin/test_admin_users.py]: does this test fail now?

File: tests/admin/test_admin_users_endpoints.py [hunk 6]
```diff
@@ -387,3 +477,257 @@ def test_put_user_username_noauth(client, db_session):
...
+
+# DELETE /users/<username> tests
+
+
+def assert_non_google_data_remained(db_session):
+    client = db_session.query(Client).filter_by(client_id="dclientid").all()
+    assert len(client) == 1
+    group = db_session.query(Group).filter_by(id=4240).all()
+    assert len(group) == 1
+    usr_grp = db_session.query(UserToGroup).filter_by(user_id=4242, group_id=4240).all()
```
Comment (Avantol13) [inline tests/admin/test_admin_users_endpoints.py]: Can the fixtures (like `load_google_specific_user_data`) return a dict with all these magic numbers and you can check the dict here ? `some_dict["userA_id"] == 4242` etc. this is just sorta hard to parse where these numbers are coming from

Comment (Avantol13) [reply tests/admin/test_admin_users_endpoints.py]: I know it's a small, annoying thing to request, but I really think it would help readability in these assert functions

File: tests/admin/test_admin_users_endpoints.py [hunk 7]
```diff
@@ -387,3 +477,257 @@ def test_put_user_username_noauth(client, db_session):
...
+        "i am not an empty dict"
+    )
+    cloud_manager.return_value.__enter__.return_value.delete_group.return_value = {}
+
+    r = client.delete(
+        "/admin/users/test_user_d",
+        headers={"Authorization": "Bearer " + encoded_admin_jwt},
+    )
+
+    assert r.status_code == 503
```
Comment (Avantol13) [inline tests/admin/test_admin_users_endpoints.py]: is it always a `503` if google api call fails? or is there a situation where a different status could happen?

**Comment by Avantol13** [comment]
> ### New Features
> * delete cascade
> * delete user and google things

@vpsx  can we change bullets to be a bit more descriptive? I know I told you at some point to be less descriptive :sweat_smile: so just need a happy medium. maybe `allow cascade deletion of related db entries` & `delete user endpoint to remove user and user-related items from db (and Google if applicable)`?

mostly I'm being picky b/c now we actually create release notes and make them available to 
users and sponsors so just want to make sure it's descriptive enough to say exactly what changed but short enough to fit as a bullet

File: fence/resources/admin/admin_users.py (hunk 1 above)
Comment (vpsx) [reply fence/resources/admin/admin_users.py]: Ah, sorry, this comment is from before the refactor and referred to `google_proxy_group_g` in `delete_user()`. The g just means google, as in "found from google, not from fence db". 

More background: `delete_user()` checks fence db for a proxy group, but as we discussed, it is possible that there's a proxy group in google without an entry in fence db. So to be on the safe side, we decided that if a proxy group is not found in fence, then we double check like this... https://github.com/uc-cdis/fence/blob/1a03c4d9eff4a9930f7021b93b74de15c68f30dd/fence/resources/admin/admin_users.py#L343-L353 

I can rename these to `google_proxy_group_from_google` or `...from_fence`? Or I can add a comment somewhere making this more explicit?

Comment (Avantol13) [reply fence/resources/admin/admin_users.py]: ah, I see now. I've been using the pattern of calling an entry in the fence db `*_db_entry` I think, but even `from_fence` instead of just `f` I think would be better :+1:

Comment (vpsx) [reply fence/resources/admin/admin_users.py]: OK! I'll use `from_fence` and `from_google` since otherwise a whole bunch of db entries would also have to be renamed to stay consistent? Also since the thing to pay attention to is really the "from google" part

File: tests/admin/test_admin_users_endpoints.py (hunk 6 above)
Comment (vpsx) [reply tests/admin/test_admin_users_endpoints.py]: THIS MAKES SO MUCH SENSE 🤦‍♂️

File: tests/admin/test_admin_users_endpoints.py (hunk 7 above)
Comment (vpsx) [reply tests/admin/test_admin_users_endpoints.py]: (For posterity, post-discussion post-edit tldr: Yes. Cirrus might throw a cirrus error; fence will catch and throw 503 instead, but still log the cirrus error.)

File: tests/admin/test_admin_users.py (hunk 5 above)
Comment (vpsx) [reply tests/admin/test_admin_users.py]: (Discussed offline; tldr for posterity: mocked out gcm here, and in userdatamodel [fixed an old NoneType error](https://github.com/uc-cdis/userdatamodel/pull/48/commits/d579f392c7f4a2cc6c36877e1559d99158b15957) and [removed some old relationships](https://github.com/uc-cdis/userdatamodel/pull/48/commits/7f391e0767adbbff7f273c88d65ba9745780ec62). Re-enabled test yay)

File: fence/resources/admin/admin_users.py [hunk 8]
```diff
@@ -150,19 +166,260 @@ def add_user_to_projects(current_session, username, projects=None):
...
+    )
+    try:
+        r = gcm.delete_group(gpg_email)
+    except Exception as e:
+        logger.exception(e)
+        raise UnavailableError(
+            "Error: Google unable to delete proxy group {}. Aborting".format(gpg_email)
+        )
+
+    if r == {}:
```
Comment (philloooo) [inline fence/resources/admin/admin_users.py]: in general, if the else statement just exit, put the short branch in front?
```
if r != {}:
  rage quit
go_on_without_indent
```
